File: solutions/sonnet-1/sonnet/ruby/brackets.rb
pairs = { ')' => '(', ']' => '[', '}' => '{' }
opens = pairs.values

STDIN.each_line do |line|
  line = line.chomp
  stack = []
  balanced = true

  line.each_char do |ch|
    if opens.include?(ch)
      stack.push(ch)
    elsif pairs.key?(ch)
      if stack.pop != pairs[ch]
        balanced = false
        break
      end
    end
  end

  balanced = false unless stack.empty?
  puts(balanced ? 'yes' : 'no')
end

File: solutions/sonnet-1/sonnet/perl/brackets.pl
use strict;
use warnings;

my %pairs = (
    ')' => '(',
    ']' => '[',
    '}' => '{',
);

while (my $line = <STDIN>) {
    chomp $line;
    my @stack;
    my $ok = 1;
    for my $ch (split //, $line) {
        if ($ch eq '(' || $ch eq '[' || $ch eq '{') {
            push @stack, $ch;
        } elsif (exists $pairs{$ch}) {
            if (@stack && $stack[-1] eq $pairs{$ch}) {
                pop @stack;
            } else {
                $ok = 0;
                last;
            }
        }
    }
    $ok = 0 if @stack;
    print $ok ? "yes\n" : "no\n";
}
